### PR TITLE
Team JZ: Schematic & PCB layout

### PR DESCRIPTION
**Pull request by JackToaster**
I've finished the schematic and created a PCB layout This PR doesn't include any firmware, it only has the generated code from CubeMX.
My board has 2 channels each of analog->CAN and CAN->analog (Using the microcontroller's built-in DAC and an LM358 dual op-amp for a 5v output). There are also optional CAN termination resistors.
This is the most complex PCB I've designed and also the first time I've designed anything with analog functionality, so please let me know if there's anything I need to add/remove/change!

### CODE REVIEW

File: Firmware/Core/Inc/main.h
Comment (izangell) [inline Firmware/Core/Inc/main.h]: Nit: Make sure firmware is under `/firmware/` (all lowercase), I'm trying to standardize this so I can implement the same workflow across all repos. My fault for not updating the template repo correctly.

**Comment by izangell** [comment]
PCB design looks excellent to my eye, great work!

**Comment by zainahm3d** [comment]
Since I can't make a normal style review with comments pointing at lines, I'm just going to throw some comments in your direction @JackToaster 

Overall, this looks fantastic and would definitely work well as it sits. Most of my comments are going to be style, not super technical things because you have those down. 

I'll split things into individual comments so you can mark them as resolved one by one. That way changes will be more traceable.

**Comment by zainahm3d** [comment]
a2c.schdoc
First of all, using a diode for reverse polarity protection is a good idea, but you end up with a voltage drop across the diode which ends up burning away power unnecessarily. You can choose to leave this in as it is, or replace with a dedicated IC that handles this for you. The difference is that the IC will use a low RDSon MOSFET which will burn away nearly no power. These chips are called "ideal diodes" and act like a perfect diode: https://www.maximintegrated.com/en/products/analog/amplifiers/MAX40200.html

Some good reading: https://www.analog.com/en/technical-articles/low-iq-ideal-diode-controller-with-reverse-input-protection.html

**Comment by zainahm3d** [comment]
a2c.schdoc
The SWD connector is a great idea, but it's 100% worth using a standardized TAG connect (preferred but you need to buy a 40$ connector once - the team should do this), or a cortex debug header (standardized) https://www.adafruit.com/product/752

I understand that this will add some size to the board, but for dev purposes it's worth the extra work. With a TAG connect, you don't even need to add a part to the PCB: https://www.tag-connect.com/product/tc2030-ctx-nl-6-pin-no-legs-cable-with-10-pin-micro-connector-for-cortex-processors?gclid=CjwKCAjw2vOLBhBPEiwAjEeK9rAntyUbUX3-w2Cl3Nyb_2YQl64srk7_FVmCzoFgiYZyZkApvGQIGRoCrKQQAvD_BwE

Either will allow you to plug directly into a JLink device with no chance of messing up plug orientation.

**Comment by zainahm3d** [comment]
a2c.schdoc
I presume the CANH/CANL connections are doubled up on J1 for easily paralleling more boards in. Great idea.

**Comment by zainahm3d** [comment]
a2c.schdoc
It's worth connecting the `mnt1` and `mnt2` pins of `J1` to ground, and having vias very close by. This will mechanically connect them to the inner layers of the board and prevent the connector from peeling off. Ditto for `J2`.

**Comment by zainahm3d** [comment]
a2c.schdoc 
I would recommend using a flag for `VBAT` instead of a net label for readability. Alternatively (preferred) go ahead and directly connect the reverse polarity protection circuit directly to `VBAT`, then it's easier to read the schematic. 
image

**Comment by zainahm3d** [comment]
a2c.schdoc
Nitpick: rotate the `3V3` flag vertically please. Same for `J2`. 
image

**Comment by zainahm3d** [comment]
STM32.schdoc
Having status LEDs is a great idea but it is currently at the cost of space. Switching to a 3 in 1 RGB LED would save a large amount of space. Additionally, I am assuming you will only illuminate one color at a time, so you can drop two of the resistors and just have common on the common lead. Yellow can be swapped for blue in an RGB config, realistically you can replace yellow with blinking green or the like. 

Also, standard LED designator is `L` instead of `DS`.

**Comment by zainahm3d** [comment]
STM32.schdoc
You may want to include a small ceramic cap on `ADC1` and `ADC2` case you need to smooth the signal in the future. Set is as a "do not populate" and then add in if needed.

**Comment by zainahm3d** [comment]
STM32.schdoc
It's a good rule of thumb to have a 100nF cap per microcontroller power lead. Have 2 of them on the 3V3 rail for the two VDD pins.

**Comment by zainahm3d** [comment]
STM32.schdoc
This is up to you but the noise filtering for the VDDA rail is likely unnecessary in this application since we don't care about super high precision. Realistically if you needed a bit more accuracy you could just oversample or average ADC readings since the CAN frame rate won't be super high compared to the ADC sample rate. Up you you if you want to keep it in place, I have no preference.

**Comment by zainahm3d** [comment]
STM32.schdoc
I'd 100% remove the reset button and just tie the `RST` line of the micro to 3V3. In an application like this you wouldn't just reset a single daughterboard, you'd flip the master kill on the car. During debugging, the JLink can remotely reset the device by setting a resister in the ARM DAP. You can tie the `RST` line to the `RESET` pin of the SWD connector, as the JLink can also drive the reset line for you. If you choose to do this make sure you have a pull up on the pin.

**Comment by zainahm3d** [comment]
STM32.schdoc
Add some TVS diodes to the `ADC_Vin` net. If there's a spike on the input there's currently no protection on the micro pin. Additionally, add a capacitor if you'd like.

**Comment by zainahm3d** [comment]
CanTransceiver.schdoc
I've gotten away with tying the RS line of the transceiver straight to ground. You can drop the resistor it if you want, no preference.

**Comment by zainahm3d** [comment]
PCB Layout
In general, the rule is to have 1 ground/power via per ground/power pad on each IC. Realistically, this isn't always possible. I see a few locations on the layout where you can add some more vias and more tightly couple the layers together. Here's an example: 
image

**Comment by zainahm3d** [comment]
PCB Layout: 
I see you're using 0402 parts. Nothing wrong with that, just be careful when ordering your stencils and opt for a thinner sheet to reduce the amount of solder used. This will also help with the STM32 pins.

**Comment by zainahm3d** [comment]
PCB Layout: 
Since this is a super tight board, I would personally use one of the assembly layers to nicely label each part, and then print out when assembling for reference. The designators can then be removed from the silk since they're too small to read anyways (most of them are only slightly wider than an 0402 part, and will not be printed perfectly by a cheap fab and will probably be unreadable anyways).

**Comment by zainahm3d** [comment]
PCB Layout: 
`R18` and `R19` are kinda huge, along with `D1` and `D2`

**Comment by zainahm3d** [comment]
PCB Layout: 
Nitpick: My imaginary application of these would prefer right angle connectors for `J1` and `J2`. This way the board can lay nicely inline with the sensor wiring.

**Comment by zainahm3d** [comment]
PCB Layout: 
Rerouting these ground lines to connect to the via just on the right for `C2` would allow you to clean up `Aout_2` and `NetR8_1`'s routing
image

**Comment by zainahm3d** [comment]
PCB Layout: 
This is another example of where there should just be two different ground vias instead of a long ground trace. 
image

**Comment by zainahm3d** [comment]
PCB Layout: 
I don't love the `SWCLK` and `SWDIO` routing here. These are likely the highest frequency signals on this board and they are also ones with some of the largest loop area with a lot of inductance. This could potentially become an issue with high frequency debug signals (during live graphing in Ozone for example). Not critical but would be nice to clean up when implementing a more standardized debug header.

**Comment by zainahm3d** [comment]
@JackToaster Overall, fantastic work. Please don't think the flood of comments are a bad thing. These are small changes that will take it from a great design to a production ready version that can be used for years without any changes.

**Comment by zainahm3d** [comment]
Sidenote: I didn't check your LED currents, your op amp gains, or your STM32 pinout selection. I trust that you can check this yourself and handle it.

**Comment by JackToaster** [comment]
@zainahm3d I made some changes to address some of your comments:

- The reverse polarity protection diode isn't ideal, but since it's a fairly low forward voltage diode and this board shouldn't draw much current I decided to leave it.
- I replaced the SWD connector with the tag connector you recommended (And re-routed the SWCLK and SWDIO lines so that they don't make a lap around the entire board)
- I replaced the connectors with a right-angle variant and connected the mount pins to ground with vias
- I fixed the orientations of flags & replaced the VBAT net label with a flag
- Switched the 3 leds with one RGB led (and a common resistor)
- Added optional capacitors to analog inputs
- Added a TVS diode to ADC-Vin (This should also protect the voltage regulator which is a big plus)
- Shrunk the overly large resistors (but left the big diodes, since I couldn't find smaller TVS diodes with similar performance)
- Removed the reset button and the resistor from the RS line on the voltage regulator
- The board is now a few millimeters smaller and a bit less tightly packed since there's fewer components
- Removed component designators from silkscreen layer & added them to the top assembly layer (I'm not 100% sure that I did this correctly)

Since the PCB was pretty heavily rearranged, your specific suggestions on routing don't apply anymore, but I tried to follow your routing advice as much as possible.

**Comment by zainahm3d** [comment]
Fantastic, now the board looks nice and compact with minimal components. 

One note that I did not make clear: the specific tag connect footprint you are using has alignment holes, but no holes for the snap-in plastic legs. This means that you can use it to program, but it will not stay connected on its own (someone has to hold the tag connect in place). Two choices here for dev work - you can either get a backing plate to put on the other side of the board to give the connector something to snap in to, or use a different variant of the connector with plastic legs that snap in (this will take a little more board space). Notice how this plug has the plastic legs: https://www.tag-connect.com/product/tc2030-ctx-6-pin-cable-for-arm-cortex?gclid=Cj0KCQiA-K2MBhC-ARIsAMtLKRu8K0I0OCjVNyFAUwiSloxU4Zdsc6IPMTHCvfce63hqesVpc2pnG6IaAiOaEALw_wcB My fault for not sending you a link with legs, sorry. 

Looking at the top assembly layer, I only see a few designators on it. Not a big deal and it shouldn't stop you from ordering parts for the board. 
image

One more thing: it's worth connecting the `SWO` pin from the STM32 to the Tag Connect. This way you can easily profile code and have another route for debug prints. 

Apart from me leading you astray on the connector this is great. Approving for now, make changes if you wish.

**Review by zainahm3d — DISMISSED** [review]

**Comment by JackToaster** [comment]
I fixed the assembly layer and added the SWO pin to the connector. There isn't room for the connector with legs unless I rearrange the whole board again, so we might be better off getting a backing plate to put on the other side of the board to give the connector something to snap in to.
image

**Comment by zainahm3d** [comment]
Took a quick look, I think it's ready to go.

**Review by zainahm3d — DISMISSED** [review]
Go ahead and generate gerbers and squash the commits down to one feature (if that's how @trisangell want it), then I'll reapprove for main.

**Comment by izangell** [comment]
@zainahm3d For reference, I have everything set to squash and merge by default. The final approver should technically be the one to press the merge button. 

@JackToaster Go ahead and re-request a review once you generate gerbers, I will enable auto-merge

**Review by izangell — APPROVED** [review]

**Pull request merged**
